aidial_assistant/chain/history.py

```diff
@@ -91,3 +123,63 @@ def to_best_effort_messages(
         )
 
         return messages
+
+    async def trim(
+        self, max_prompt_tokens: int, model_client: ModelClient
+    ) -> "History":
+        extra_results_callback = ModelExtraResultsCallback()
+        stream = model_client.agenerate(
+            self.to_protocol_messages(),
+            extra_results_callback,
+            max_prompt_tokens=max_prompt_tokens,
+            max_tokens=1,
+        )
+        try:
+            async for _ in stream:
+                pass
+        except ReasonLengthException:
+            # Expected for max_tokens=1
+            pass
+
+        if extra_results_callback.discarded_messages:
+            return History(
+                assistant_system_message_template=self.assistant_system_message_template,
+                best_effort_template=self.best_effort_template,
+                scoped_messages=self._skip_messages(
+                    extra_results_callback.discarded_messages
+                ),
+            )
+
+        return self
+
+    def _skip_messages(self, message_count: int) -> list[ScopedMessage]:
+        messages = []
+        current_message = self.scoped_messages[0]
+        message_iterator = iter(self.scoped_messages)
+        for _ in range(message_count):
+            current_message = next(message_iterator)
+            while current_message.message.role == Role.SYSTEM:
+                # System messages should be kept in the history
+                messages.append(current_message)
+                current_message = next(message_iterator)
+
+        if current_message.scope == MessageScope.INTERNAL:
+            while current_message.scope == MessageScope.INTERNAL:
+                current_message = next(message_iterator)
+
+            # Internal messages (i.e. addon requests/responses) are always followed by an assistant reply
+            assert (
+                current_message.message.role == Role.ASSISTANT
+            ), "Internal messages must be followed by an assistant reply."
+
+        remaining_messages = list(message_iterator)
+        assert (
+            len(remaining_messages) > 0
+        ), "No user messages left after history truncation."
```

Comment:
Let's use the same terminology across the whole project. Either "trimming" or "truncation". Currently I see both used in various places.

Reply:
Renamed trimming to truncation.